src/input/mpegts/iptv/iptv_rtsp.c

```diff
@@ -270,26 +382,176 @@ iptv_rtp_header_callback ( iptv_mux_t *im, uint8_t *rtp, int len )
 static ssize_t
 iptv_rtsp_read ( iptv_input_t *mi, iptv_mux_t *im )
 {
-  rtsp_priv_t *rp = im->im_data;
-  udp_multirecv_t *um = &rp->um;
   ssize_t r;
-  uint8_t buf[1500];
-
-  /* RTCP - ignore all incoming packets for now */
-  do {
-    r = recv(im->mm_iptv_fd2, buf, sizeof(buf), MSG_DONTWAIT);
-  } while (r > 0);
-
-  r = iptv_rtp_read(im, um, iptv_rtp_header_callback);
+  if (im->mm_iptv_send_reports) {
+    uint8_t buf[1500];
+    /* RTCP - ignore all incoming packets for now */
+    do {
+      r = recv(im->im_rtcp_info.connection_fd, buf, sizeof(buf), MSG_DONTWAIT);
+    } while (r > 0);
+    r = iptv_rtp_read(im, iptv_rtp_header_callback);
+  } else
+    r = iptv_rtp_read(im, NULL);
   if (r < 0 && ERRNO_AGAIN(errno))
     r = 0;
   return r;
 }
 
 /*
- * Initialise RTSP handler
+ * Send the status message
  */
+#if ENABLE_TIMESHIFT
+static void rtsp_timeshift_fill_status(rtsp_st_t *ts, rtsp_priv_t *rp,
+    timeshift_status_t *status) {
+  int64_t start, end, current;
+
+  if (rp == NULL) {
+    start = 0;
+    end = 3600;
+    current = 0;
+  } else {
+    start = 0;
+    end = rp->range_end - rp->range_start;
+    current = rp->position - rp->range_start;
+  }
+  status->full = 0;
+
+  tvhdebug(LS_TIMESHIFT,
+      "remote ts status start %"PRId64" end %"PRId64 " current %"PRId64, start, end,
+      current);
+
+  status->shift = ts_rescale_inv(current, 1);
+  status->pts_start = ts_rescale_inv(start, 1);
+  status->pts_end = ts_rescale_inv(end, 1);
+}
+
+static void rtsp_timeshift_status
+  ( rtsp_st_t *pd, rtsp_priv_t *rp )
+{
+  streaming_message_t *tsm, *tsm2;
+  timeshift_status_t *status;
+
+  status = calloc(1, sizeof(timeshift_status_t));
+  rtsp_timeshift_fill_status(pd, rp, status);
+  tsm = streaming_msg_create_data(SMT_TIMESHIFT_STATUS, status);
+  tsm2 = streaming_msg_clone(tsm);
+  streaming_target_deliver2(pd->output, tsm);
+  streaming_target_deliver2(pd->tsfix, tsm2);
+}
+
+void *rtsp_status_thread(void *p) {
+  int64_t mono_now, mono_last_status = 0;
+  rtsp_st_t *pd = p;
+  rtsp_priv_t *rp;
+
+  while (pd->run) {
+    mono_now  = getfastmonoclock();
+    if(pd->im == NULL)
+      continue;
+    rp = (rtsp_priv_t*) pd->im->im_data;
+    if(rp == NULL || !pd->rtsp_input_start)
+      continue;
+    if (mono_now >= (mono_last_status + sec2mono(1))) {
+      // In case no buffer updates available assume the buffer is being filled
+      if(rp->hc && rp->hc->hc_rtsp_keep_alive_cmd != RTSP_CMD_DESCRIBE)
+        rp->range_end++;
+      rtsp_timeshift_status(pd, rp);
+      mono_last_status = mono_now;
+    }
+  }
+  return NULL;
+}
+
+static void rtsp_input(void *opaque, streaming_message_t *sm) {
+  int type = sm->sm_type;
+  rtsp_st_t *pd = (rtsp_st_t*) opaque;
+  iptv_mux_t *mux;
+  streaming_skip_t *data;
+  rtsp_priv_t *rp;
+
+  if (pd == NULL || (sm->sm_s == NULL && pd->im == NULL)) {
+    streaming_target_deliver2(pd->output, sm);
```

Comment:
You check if `pd == NULL` above and here you refer to `pd->output` if `pd == NULL`. That will cause a crash if pd is NULL.